### PR TITLE
ROC-7391 Update contact details

### PR DESCRIPTION
**Pull request by Valkrys**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/ROC-7391

### Change description ###
Handle contact details changes by case workers  and notify parties

### CODE REVIEW

**Comment by timja** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/util/MapperUtil.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/events/GeneralLetterReadyToPrintEvent.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterCallbackHandler.java  1
- src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/generalletter/GeneralLetterCallbackHandlerTest.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java  2
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/ChangeContactDetailsNotificationService.java  15
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBodyMapper.java  4
- src/integrationTest/java/uk/gov/hmcts/cmc/claimstore/generalletter/GeneralLetterServiceTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/HMCTS/cmc-claim-store/pullRequest?prid=5210122&bid=17085394)

**Comment by timja** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/uk/gov/hmcts/cmc/claimstore/events/GeneralLetterReadyToPrintEvent.java  1
- ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/util/MapperUtil.java  1
- src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterCallbackHandlerTest.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBodyMapper.java  4
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/ChangeContactDetailsNotificationService.java  15
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java  3
- src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterServiceTest.java  1
- src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterCallbackHandler.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/HMCTS/cmc-claim-store/pullRequest?prid=5210122&bid=17085394)

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/ChangeContactDetailsNotificationService.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+                .builder()
+                .data(caseDetailsConverter.convertToMap(updatedCase))
+                .build();
+        } else {
+            return AboutToStartOrSubmitCallbackResponse
+                .builder()
+                .errors(Collections.singletonList(ERROR_MESSAGE))
+                .build();
+        }
+    }
```
Comment (daveoh) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/ChangeContactDetailsNotificationService.java:79]: When returning an exception this is a better pattern to follow as it avoids nested ifs in the future and puts the happy path return object at the end of the method

```suggestion
        if (errors) {
              return AboutToStartOrSubmitCallbackResponse
                .builder()
                .errors(Collections.singletonList(ERROR_MESSAGE))
                .build();
        }
        
        logger.info("Change Contact Details: Email was sent, case is being updated");
        CCDCase updatedCase = ccdCase.toBuilder()
              .contactChangeParty(null)
              .contactChangeContent(null)
               generalLetterContent(null)
               build();
          return AboutToStartOrSubmitCallbackResponse
              .builder()
              .data(caseDetailsConverter.convertToMap(updatedCase))
              .build();
    }
```

File: src/main/java/uk/gov/hmcts/cmc/claimstore/stereotypes/EmailContentProvider.java [hunk 2]
```diff
@@ -18,7 +18,7 @@
     EmailContent createContent(T input);
 
     default String evaluateTemplate(String template, Map<String, Object> input) {
-        return getTemplateService().evaluate(template, input).trim();
+        return getTemplateService().evaluate(template, input);
```
Comment (daveoh) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/stereotypes/EmailContentProvider.java:21]: Do we not want to trim this anymore?

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/ChangeContactDetailsCallbackHandler.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import uk.gov.hmcts.cmc.claimstore.services.ccd.callbacks.CallbackType;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import static uk.gov.hmcts.cmc.ccd.domain.CaseEvent.CHANGE_CONTACT_DETAILS;
+import static uk.gov.hmcts.cmc.claimstore.services.ccd.Role.CASEWORKER;
+
+@Service
```
Comment (akritis22) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/ChangeContactDetailsCallbackHandler.java:21]: Please put this behind feature toggle

Comment (akritis22) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/ChangeContactDetailsCallbackHandler.java:21]: Please put this behind feature toggle

Comment (daveoh) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/ChangeContactDetailsCallbackHandler.java:21]: I am just getting confirmation from Brian whether this is needed or not. The ticket is not clear

Comment (dharmendrak) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/ChangeContactDetailsCallbackHandler.java:21]: put up the feature toggle in it

File: src/main/java/uk/gov/hmcts/cmc/claimstore/stereotypes/EmailContentProvider.java (hunk 2 above)
Comment (dharmendrak) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/stereotypes/EmailContentProvider.java]: reverted

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/ChangeContactDetailsNotificationService.java (hunk 1 above)
Comment (dharmendrak) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/ChangeContactDetailsNotificationService.java]: I have removed the usage off errors flag

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/ChangeContactDetailsNotificationService.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+    private void notifyDefendant(Claim claim) {
+        notificationService.sendMail(
+            claim.getDefendantEmail(),
+            notificationsProperties.getTemplates().getEmail().getClaimantContactDetailsChanged(),
+            aggregateParams(claim),
+            NotificationReferenceBuilder.ContactDetailsChanged
+                .referenceForDefendant(claim.getReferenceNumber(), "defendant")
+        );
+    }
```
[the remark this answers is not in the document]
Comment (dharmendrak) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/ChangeContactDetailsNotificationService.java]: done

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/notifications/NotificationReferenceBuilder.java [hunk 5]
```diff
@@ -254,4 +254,20 @@ public static String forMediationUnsuccessfulOfflineDQ(String claimReferenceNumb
...
+            // do not instantiate
+        }
+
+        public static String referenceForClaimant(String claimReferenceNumber, String otherParty) {
+            return reference(TEMPLATE, otherParty.toLowerCase(), claimReferenceNumber);
+        }
+
+        public static String referenceForDefendant(String claimReferenceNumber, String otherParty) {
+            return reference(TEMPLATE, otherParty.toLowerCase(), claimReferenceNumber);
+        }
```
[the remark this answers is not in the document]
Comment (dharmendrak) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/notifications/NotificationReferenceBuilder.java]: done

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/ChangeContactDetailsNotificationService.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+@Service
+public class ChangeContactDetailsNotificationService {
+
+    private static final String ERROR_MESSAGE =
+        "There was a technical problem. Nothing has been sent. You need to try again.";
+    private final NotificationService notificationService;
+    private final NotificationsProperties notificationsProperties;
+    private final CaseDetailsConverter caseDetailsConverter;
+    private static final Logger logger = LoggerFactory.getLogger(LoggerHandler.class);
```
Comment (daveoh) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/ChangeContactDetailsNotificationService.java:37]: This needs to change to 
```suggestion
    private static final Logger logger = LoggerFactory.getLogger(ChangeContactDetailsNotificationService.class);
```

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/ChangeContactDetailsPostProcessor.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+import java.util.Collections;
+
+import static uk.gov.hmcts.cmc.ccd.domain.CCDContactPartyType.CLAIMANT;
+import static uk.gov.hmcts.cmc.claimstore.services.ccd.callbacks.CallbackParams.Params.BEARER_TOKEN;
+
+@Service
+@ConditionalOnProperty(prefix = "doc_assembly", name = "url")
+public class ChangeContactDetailsPostProcessor {
+
+    private static final Logger logger = LoggerFactory.getLogger(LoggerHandler.class);
```
Comment (daveoh) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/ChangeContactDetailsPostProcessor.java:38]: This needs to change to 
```suggestion
    private static final Logger logger = LoggerFactory.getLogger(ChangeContactDetailsPostProcessor.class);
```

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/ChangeContactDetailsNotificationService.java (hunk 6 above)
Comment (dharmendrak) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/ChangeContactDetailsNotificationService.java]: done

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/ChangeContactDetailsPostProcessor.java (hunk 7 above)
Comment (dharmendrak) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/ChangeContactDetailsPostProcessor.java]: done

**Pull request merged**